src/DataStructures/Tensor/Expressions/KroneckerDeltaAsExpression.hpp

```diff
@@ -0,0 +1,174 @@
+  /// point of a leg along the primary path
+  static constexpr bool primary_subtree_contains_primary_start =
+      is_primary_start;
+
+  /// \brief Construct a `KroneckerDeltaAsExpression` from a `KroneckerDelta`
+  ///
+  /// \param k the `KroneckerDelta` to represent as a `TensorExpression`
+  KroneckerDeltaAsExpression(const K& k) : k_(&k) {}
+  ~KroneckerDeltaAsExpression() override = default;
+
+  // This expression does not represent a `Tensor`, nor does it have any
+  // children, so we should never need to assert that the LHS `Tensor` is not
+  // equal to the `KroneckerDelta` stored by this expression
+  template <typename LhsTensor>
+  void assert_lhs_tensor_not_in_rhs_expression(
+      const gsl::not_null<LhsTensor*>) const = delete;
+  // This expression does not represent a `Tensor`, nor does it have any
+  // children, so we should never need to assert that the LHS `Tensor` is not
+  // equal to the `KroneckerDelta` stored by this expression
+  template <typename LhsTensorIndices, typename LhsTensor>
+  void assert_lhs_tensorindices_same_in_rhs(
+      const gsl::not_null<LhsTensor*> lhs_tensor) const = delete;
+
+  /// \brief Returns the value of the contained Kronecker delta's multi-index
+  ///
+  /// \param multi_index the multi-index of the component to retrieve
+  /// \return the value of the component at `multi_index` in the Kronecker delta
+  constexpr SPECTRE_ALWAYS_INLINE double get(
+      const std::array<size_t, num_tensor_indices>& multi_index) const {
+    if (gsl::at(multi_index, 0) != gsl::at(multi_index, 1)) {
+      return 0.0;
+    } else {
+      return 1.0;
+    }
+  }
+
+  /// \brief Returns the value of the contained Kronecker delta's multi-index
+  ///
+  /// \param multi_index the multi-index of the component to retrieve
+  /// \return the value of the component at `multi_index` in the Kronecker delta
+  template <typename ResultType>
+  constexpr SPECTRE_ALWAYS_INLINE double get_primary(
+      const ResultType& /*result_component*/,
+      const std::array<size_t, num_tensor_indices>& multi_index) const {
+    if (gsl::at(multi_index, 0) != gsl::at(multi_index, 1)) {
+      return 0.0;
+    } else {
+      return 1.0;
+    }
+  }
+
+  // This expression is a leaf but does not store any information related to the
+  // sizing of the tensor components because it does not represent an actual
+  // `Tensor`. Therefore, this expression should never be used to initialize a
+  // LHS result tensor component. We would run into trouble if e.g. the tensor
+  // components in the equations are `DataVector`s, but then we initialize a LHS
+  // component using a Kronecker delta element stored in this leaf expression on
+  // the primary path.
+  template <typename ResultType>
+  void evaluate_primary_subtree(
+      ResultType&,
+      const std::array<size_t, num_tensor_indices>&) const = delete;
+
+ private:
+  /// Kronecker delta represented by this expression
+  const K* k_ = nullptr;
```

Comment:
Can't you just store a `KroneckerDelta` object instead of a pointer? It's an empty object and so the same size as a pointer (or smaller? 1 byte, I think). Why even bother storing it? It doesn't look like you use it.